crypto/rand/drbg_lib.c

```diff
@@ -33,9 +28,79 @@ static const char ossl_pers_string[] = "OpenSSL NIST SP 800-90A DRBG";
  * a much bigger deal than just re-setting an allocated resource.)
  */
 
+/*
+ * The three shared DRBGs
+ *
+ * There are three shared DRBGs (master, public and private), which are
+ * accessed concurrently by all threads.
+ *
+ * The master DRBG
+ *
+ * Not used directly by the application, only for reseeding the two other
+ * DRBGs. It reseeds itself by pulling either randomness from os entropy
+ * sources or by consuming randomnes which was added by RAND_add()
+ */
+static RAND_DRBG drbg_master;
+/*
+ * The public DRBG
+ *
+ * Used by default for generating random bytes using RAND_bytes().
+ */
+static RAND_DRBG drbg_public;
+/*
+ * The private DRBG
+ *
+ * Used by default for generating private keys using RAND_priv_bytes()
+ */
+static RAND_DRBG drbg_private;
+/*
+ * In addition there are DRBGs, which are not shared, but used only by a
+ * single thread at every time, for example the DRBGs which are owned by
+ * an SSL context.
+ *
+ * This gives the following overall picture:
+ *
+ *                  <os entropy sources>
+ *                         |
+ *                         v
+ *    RAND_add() ==>    <master>          \
+ *                       /   \            | shared DRBGs
+ *                      v     v           |  (with locking)
+ *                 <public>  <private>    /
+ *                     |
+ *                     v
+ *                   <ssl>  owned by an SSL context
+ *
+ * Reseeding
+ *
+ * The master DRBG reseeds whenever
+ *
+ *  - the number of generate requests exceeds the reseed_interval,
+ *  - the time elapsed since the last reseeding exceeds the
+ *    reseed_time_interval, or
+ *  - entropy was pushed by calling RAND_add().
+ *
+ * The public and private DRBG reseed whenever
+ *
+ *  - the number of generate requests exceeds the reseed_interval, or
+ *  - the reseed_time of the master has changed ("seed propagation").
+ *
+ * The ssl DRBGs reseed whenever
+ *
+ * - the number of generate requests exceeds the reseed_interval.
+ *
+ * This implies that a reseeding of the master DRBG is automatically
+ * propagated to the two other shared DRBGs, but not to the thread or
+ * SSL specific DRBGs.
```

Comment:
So I think that means if someone using the private DRBG requests prediction resistance, all 3 would get reseeded? I would expect the private and the master to get reseeded (based on my understanding of the NIST documents), but from what I understand here the next time the public one is used it would also get reseeded. It's probably not a side effect we want.

Reply:
No, currently the prediction_resistance request is not propagated to the parent. So a generate request to the private DRBG with prediction_resistance=1 only reseeds the private DRBG, not the master DRBG (unless its reseed_count reaches the reseed_interval).


Up to now it was my understanding that the NIST standard does not mandate the prediction request to be propagated to the parent DRBG: The NIST standard only considers a single DRBG. An adversary might have gained knowledge about its internal state, but not about its parent, which is considered a black box. Here are some citations from [NIST SP 800-90Ar1](http://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf):

*Definition of Prediction Resistance (page 6)*

> An RBG provides prediction resistance relative to time T if it provides assurance that an adversary with knowledge of the state of the RBG at some time(s) prior to T (but incapable of performing work that matches the claimed security strength of the RBG) would be unable to distinguish between observations of ideal random bitstrings and (previously unseen) bitstrings output by the RBG at or subsequent to time T. In particular, an RBG whose design allows the adversary to step forward from the initially compromised RBG state(s) to obtain knowledge of subsequent RBG states and the corresponding outputs (including the RBG state and output at time T) would not provide prediction resistance relative to time T.

On page 25, it says 

> The prediction_resistance_request parameter indicates whether or not prediction resistance is to be provided during the request (i.e., whether fresh entropy is required[4]).

Chaining of DRBGs is justified by footnote[4] of page 25:

> [4] 4 Entropy input may be obtained from an entropy source or an NRBG, both of which provide fresh entropy. Entropy
input could also be obtained from a DRBG that has access to an entropy source or NRBG. The request for prediction resistance rules out the use of a DRBG that does not have access to either an entropy source or NRBG.

and in further detail by footnote [5], page 30:
> A DRBG may be reseeded by an entropy source or an NRBG, both of which provide fresh entropy. A DRBG could also be reseeded by a DRBG that has access to an entropy source or NRBG. The request for prediction resistance during reseeding rules out the use of a DRBG that does not have access to either an entropy source or NRBG. See [SP 800-90C] for further discussion. 

In our case one might argue differently, because the master DRBG resides in the same address space as public and private DRBG. An adversary that gains knowledge about the private DRBG could indeed also have knowledge about its parent. I tried to mitigate this possility by lowering the reseed intervals to keep the total "os reseed interval" at 2^24. But I would understand if someone insisted on forwarding the prediction_resistance request to the parent. In this case one would indeed get the Domino effect you mentioned. It's the usual security vs. performance tradeoff. As such, it's a political question and should probably be decided the OMC members. I am a little in favor of my choice, but would also follow an majority vote.

Reply:
I have no idea how you can interpret that text differently. There needs to be a way to get fresh entropy from somewhere. A DRBG does not provide fresh entropy, you need something that directly talks to the entropy source to get such fresh entropy and that's the master DRBG.

Reply:
My understanding of this standard was that the request for new entropy should propagate along the chain until some fresh entropy had been gathered from a non-DRBG source.

I.e. the private DRBG should cause the master DRBG to reseed before reseeding itself.

Reply:
I disagree, and it is not a matter of interpretation, it is written down in the standard:

- Requesting prediction resistance simply means requesting a reseeding of the DRBG before generating random output. There is no difference between a "normal" reseeding and a reseeding triggered by a prediction resistance request.
- The NIST standard explicitely allows reseeding from a DRBG which is connected to an approved entropy source. There is no word about a need to reseed the parent immediately before obtaining its random output.

I you are questioning the fact that a DRBG can safely be reseeded from the master DRBG, then you question the entire construction of chaining the DRBGs to improve the security of the DRBGs in the first place.

But before going deeper into this, I would like to dispel your concerns about prediction resistance:

I agree that a kernel mode CSPRNG is more secured than a user space CSPRNG. So I completely understand if someone prefers to pull randomness from the former instead of the latter. But this is already possible to do: If the applications needs "strong" prediction resistance for the public or the private DRBG, all it has to do is to call RAND_poll() before calling RAND_bytes() resp. RAND_priv_bytes(). This will reseed the master DRBG from the os entropy sources, and the reseeding will then be propagated to the public resp. private DRBG. Alternatively, it could obtain a pointer to the master DRBG (drbg = RAND_DRBG_get0_master()) and call its generate method requesting prediction resistance.


### The facts

I was not trying to interpret the standard, I was stating the facts from the document and verifying them by giving quotations. So please tell me, where exactly I misinterpreted something in my reasoning below. And please point out the reference where it is stated that the parent DRBG's generate method SHALL be called demanding prediction resistance.

The DRBG construction is an approved method of 'stretching' the sparse resource of randomness into a longer pseudo random output. This means that the DRBG's output is computationally indistinguishable from a true random coin toss for an adversary whose computational capabilities lie within the limits assumed for the given security level (not more and not less). This holds under the assumptions that

- the DRBG was reseeded from an approved entropy source or nondeterministic RNG
- the internal state of the DRBG is unknown to the adversary

If the previous statement holds then not only the first block of 128 bits generated by the parent DRBG is computationally indistinguishable from true random output, but also the second and the third block, etc. So why should the first 128 bit block contain "more entropy" than the second? Neither of them contains any entropy because the output is a deterministic function of the internal state. All the randomness lies in the random choice of this internal state, and the fact that the internal state remains secret.

Nothing else happens when you reseed from the kernel's RNG: You will get some block of pseudo random output from a deterministic RNG and you can't even be certain that enough bits of "fresh" entropy have been added to the kernel's random pool since your last request (more on this subject below).

Calling RAND_DRBG_generate() with prediction_resistance=1 is the request to "reseed before generating output", in order to handle the case that the internal state of the DRBG might have been (partially) compromised. It triggers a "normal" reseed, without any additional constraints on the "quality" of the random input pulled from the entropy source.

The NIST standard explicitely states that instead of an approved nondeterministic RNG or entropy source it is allowed to use a DRBG which is seeded by an approved nondeterministic RNG or entropy source. Indeed, unless the state of the parent DRBG is compromised, too, or the randomness input fed into the child DRBG is intercepted by the adversary, this reseeding will immediately restore the integrity of the child DRBG.


### The myth

What is "fresh" entropy in the first place? Is it filled into the DRBG and used up by generate requests like the gasoline of an electricity generator (pun intended)? No, this is just and analogy for humans and taking this analogy too literal is the cause for the misbelief that a DRBG which has been seeded with, say, 128 bits of entropy gets drained after generating 128 bits of random output. This superstition caused generations of programmers to block on /dev/random [waiting for "fresh" entropy](https://www.2uo.de/myths-about-urandom/#low-entropy). I don't know the author of the linked article, but I agree with his arguments, in particular with the [quote of Daniel Bernstein](https://www.2uo.de/myths-about-urandom/#experts), which was taken from [this email](https://www.mail-archive.com/cryptography@randombit.net/msg04763.html)).


### Theory vs. practice

In the context of the NIST document the case is clear: it is not necessary to propagate the prediction resistance request to the master. In our case the situation is a little more complicated, because both DRBGs are located in the address space of the same process, so it is not clear whether the master DRBG's internal state can be considered secure after the child DRBG's internal state was compromised.

Whether or not the master DRBG can be considered secure depends on the threat model. If the adversary can only intercept network traffic and learn about the state of the child DRBG, but does not get to see any of the master DRBG's output, then the answer is clearly yes. (Apropos: it's probably better not to reseed the ssl DRBGs from the public DRBG, because we have no control how much output of the public DRBG will be used for other purposes and become publicly known.)

If the adversary obtained knowledge of the child DRBG's internal state by reading some portion of the process memory, then it is arguable that he might now about the internal state of the master, too. One mitigation for this threat would be to not use static variables to hold the internal state of the three DRBGs, but allocate it from the secure heap.

Anyway, as stated above: It is easy to trigger a full reseed from the os entropy sources either by calling RAND_poll() or by using a DRBG which is not chained and requesting prediction resistance.